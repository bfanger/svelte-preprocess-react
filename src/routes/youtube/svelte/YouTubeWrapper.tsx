import { reactify } from "svelte-preprocess-react";
import { Youtube } from "svelte-youtube-lite";

const YoutubeWrapper = reactify(Youtube);
export default YoutubeWrapper;
