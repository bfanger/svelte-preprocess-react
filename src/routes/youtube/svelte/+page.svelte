<script>
  import { Youtube } from "svelte-youtube-lite";
  import YouTubeWrapper from "./YouTubeWrapper";
</script>

<react:YouTubeWrapper id="AdNJ3fydeao" />
