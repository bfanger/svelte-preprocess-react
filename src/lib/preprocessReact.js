import MagicString from "magic-string";
import { compile, preprocess } from "svelte/compiler";
import detectReactVersion from "./internal/detectReactVersion.js";

/**
 * @typedef {import("svelte/compiler").PreprocessorGroup} PreprocessorGroup
 * @typedef {import("svelte/compiler").Processed} Processed
 * @typedef {import("svelte/types/compiler/interfaces").TemplateNode} TemplateNode
 */

const defaults = /** @type {const} */ ({
  react: "auto",
  ssr: true,
  errorMode: "throw",
});

/**
 * Svelte preprocessor to use convert <react:*> tags into Sveltified React components.
 *
 * Imports renderToString from 'react-dom/server' unless the `ssr` option is set to false.
 *
 * @param {{
 *   ssr?: boolean;
 *   react?: number | "auto";
 *   errorMode?: "throw" | "warn";
 *   preprocess?: PreprocessorGroup | PreprocessorGroup[];
 * }} options
 * @returns {PreprocessorGroup}
 */
export default function preprocessReact(options = {}) {
  let react = options.react ?? defaults.react;
  const ssr = options.ssr ?? defaults.ssr;
  const errorMode = options.errorMode ?? defaults.errorMode;

  return {
    async markup({ content, filename }) {
      try {
        /** @type {Processed | undefined} */
        let preprocessed;
        if (options.preprocess) {
          preprocessed = await preprocess(content, options.preprocess, {
            filename,
          });
          // eslint-disable-next-line no-param-reassign
          content = preprocessed.code;
        }

        if (react === "auto") {
          react = await detectReactVersion();
        }
        const processed = transform(content, {
          filename,
          react,
          ssr,
          errorMode,
        });
        if (!preprocessed) {
          return processed;
        }
        if (!processed.map) {
          return preprocessed;
        }
        return {
          code: processed.code,
          map: preprocessed.map ?? processed.map, // @todo apply sourcemaps
          dependencies: preprocessed.dependencies,
        };
      } catch (err) {
        if (errorMode === "throw") {
          throw err;
        }
        console.warn(err);
        return { code: content };
      }
    },
  };
}

/**
 * @param {string} content
 * @param {{
 *  filename?: string;
 *  react: number;
 *  ssr: boolean;
 *  errorMode: import("svelte/compiler").CompileOptions["errorMode"]
 * }} options
 * @returns
 */
function transform(content, options) {
  const prefix = "React$$";
  /** @type {string} */
  let portal;
  const imports = [
    `import { sveltify as ${prefix}sveltify } from "svelte-preprocess-react"`,
  ];
  if (options.react >= 18) {
    imports.push(
      `import ${prefix}ReactDOM from "react-dom/client"`,
      `import { createPortal as ${prefix}createPortal} from "react-dom"`,
    );
    portal = `${prefix}createPortal`;
  } else {
    imports.push(`import ${prefix}ReactDOM from "react-dom"`);
    portal = `${prefix}ReactDOM.createPortal`;
  }

  let renderToString = "";
  if (options.ssr) {
    imports.push(
      `import { renderToString as ${prefix}renderToString } from "react-dom/server"`,
    );
    renderToString = `, ${prefix}renderToString`;
  }

  const compiled = compile(content, {
    filename: options.filename,
    errorMode: options.errorMode,
    generate: false,
  });
  const s = new MagicString(content, { filename: options.filename });
  const components = replaceReactTags(compiled.ast.html, s);
  const aliases = Object.entries(components);

  if (aliases.length === 0) {
    return { code: content };
  }
  const script = compiled.ast.instance || compiled.ast.module;
  const wrappers = aliases
    .map(
      ([alias, expression]) =>
        `const ${alias} = ${prefix}sveltify(${expression}, ${portal}, ${prefix}ReactDOM${renderToString});`,
    )
    .join(";");

  if (!script) {
    s.prepend(`<script>\n${imports.join("; ")}\n\n${wrappers}\n</script>\n\n`);
  } else {
    /** @type {any} */
    const program = script.content;
    s.appendRight(program.end, wrappers);
    s.appendRight(program.start, `${imports.join("; ")}; `);
  }
  return {
    code: s.toString(),
    map: s.generateMap(),
  };
}

/**
 * Replace react:* tags by injecting Sveltified versions of the React components.
 *
 * @param {TemplateNode} node
 * @param {MagicString} content
 * @param {Record<string, string>} components
 */
function replaceReactTags(node, content, components = {}) {
  /* eslint-disable no-param-reassign */
  if (node.type === "Element" && node.name.startsWith("react:")) {
    /** @type {import("svelte/types/compiler/interfaces").Element} */
    const tag = /** @type {any} */ (node);
    const componentExpression = tag.name.slice(6);
    const alias = `React$${componentExpression.replace(/\./g, "$")}`;
    const tagStart = node.start;
    const tagEnd = node.end;
    const closeStart = tagEnd - tag.name.length - 3;
    const hasCloseTag =
      content.slice(closeStart, closeStart + 8) === `</react:`;
    content.overwrite(tagStart + 1, tagStart + 1 + tag.name.length, alias);
    if (hasCloseTag) {
      content.overwrite(
        closeStart + 2,
        closeStart + 2 + tag.name.length,
        alias,
      );
    }
    if (!components[alias]) {
      if (componentExpression.match(/^[a-z-]+$/)) {
        components[alias] = `"${componentExpression}"`;
      } else {
        components[alias] = componentExpression;
      }
    }
    tag.attributes.forEach((attr) => {
      if (attr.type === "EventHandler" && attr.expression !== null) {
        const event = attr;
        if (event.modifiers.length > 0) {
          throw new Error(
            "event modifier are not (yet) supported for React components",
          );
        }
        const eventStart = event.start;
        content.overwrite(
          eventStart,
          eventStart + 4,
          `on${event.name[0].toUpperCase()}`,
        );
      }
    });
    if (node.children && node.children.length > 0) {
      const isTextContent =
        node.children.filter(
          (child) => ["Text", "MustacheTag"].includes(child.type) === false,
        ).length === 0;
      /** @type {string[]} */
      const escaped = [];
      if (isTextContent) {
        // Convert text & expressions into a children prop.
        escaped.push('"');
        node.children.forEach((child) => {
          if (child.type === "Text") {
            escaped.push(
              child.data.replace(/"/g, `{'"'}`).replace(/\n/g, `{'\\n'}`),
            );
          } else if (child.type === "MustacheTag") {
            const expression = content.original.slice(child.start, child.end);
            escaped.push(expression);
          } else {
            throw new Error(`Unexpected node type:${child.type}`);
          }
        });
        escaped.push('"');
        // slot was converted to children prop
        content.appendRight(
          node.children[0].start - 1,
          ` children=${escaped.join("")} /`,
        );
        content.remove(node.children[0].start, node.end);
        return components;
      }
    }
  }
  /**
   * @param {TemplateNode} child
   */
  function processChild(child) {
    replaceReactTags(child, content, components);
  }
  // traverse children & branching blocks
  node.children?.forEach(processChild);
  node.else?.children?.forEach(processChild);
  node.pending?.children?.forEach(processChild);
  node.then?.children?.forEach(processChild);
  node.catch?.children?.forEach(processChild);
  return components;
}
