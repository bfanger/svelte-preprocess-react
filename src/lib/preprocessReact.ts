import MagicString from "magic-string";
import { compile, preprocess } from "svelte/compiler";
import type {
  CompileOptions,
  Element,
  TemplateNode,
  Transition,
} from "svelte/types/compiler/interfaces";
import type {
  PreprocessorGroup,
  Processed,
} from "svelte/types/compiler/preprocess";
import detectReactVersion from "./internal/detectReactVersion.js";

type Options = {
  react?: number | "auto";
  ssr?: boolean; // Set to false to omit renderToString from 'react-dom/server'
  errorMode?: "throw" | "warn";
  preprocess?: PreprocessorGroup | PreprocessorGroup[];
};

const defaults = {
  react: "auto",
  ssr: true,
  errorMode: "throw",
} as const;
export default function preprocessReact(
  options: Options = {},
): PreprocessorGroup {
  let react = options.react ?? defaults.react;
  const ssr = options.ssr ?? defaults.ssr;
  const errorMode = options.errorMode ?? defaults.errorMode;

  return {
    async markup({ content, filename }) {
      try {
        let preprocessed: Processed | undefined;
        if (options.preprocess) {
          preprocessed = await preprocess(content, options.preprocess, {
            filename,
          });
          // eslint-disable-next-line no-param-reassign
          content = preprocessed.code;
        }

        if (react === "auto") {
          react = await detectReactVersion();
        }
        const processed = transform(content, {
          filename,
          react,
          ssr,
          errorMode,
        });
        if (!preprocessed) {
          return processed;
        }
        if (!processed.map) {
          return preprocessed;
        }
        return {
          code: processed.code,
          map: preprocessed.map ?? processed.map, // @todo apply sourcemaps
          dependencies: preprocessed.dependencies,
        };
      } catch (err) {
        if (errorMode === "throw") {
          throw err;
        }
        console.warn(err);
        return { code: content };
      }
    },
  };
}
type TransformOptions = {
  filename?: string;
  react: number;
  ssr: boolean;
  errorMode: CompileOptions["errorMode"];
};
function transform(content: string, options: TransformOptions) {
  const prefix = "React$$";
  let portal: string;
  const imports = [
    `import ${prefix}sveltify from "svelte-preprocess-react/sveltify"`,
  ];
  if (options.react >= 18) {
    imports.push(
      `import ${prefix}ReactDOM from "react-dom/client"`,
      `import { createPortal as ${prefix}createPortal} from "react-dom"`,
    );
    portal = `${prefix}createPortal`;
  } else {
    imports.push(`import ${prefix}ReactDOM from "react-dom"`);
    portal = `${prefix}ReactDOM.createPortal`;
  }

  let renderToString = "";
  if (options.ssr) {
    imports.push(
      `import { renderToString as ${prefix}renderToString } from "react-dom/server"`,
    );
    renderToString = `, ${prefix}renderToString`;
  }

  const compiled = compile(content, {
    filename: options.filename,
    errorMode: options.errorMode,
    generate: false,
  });
  const s = new MagicString(content, { filename: options.filename });
  const components = replaceReactTags(compiled.ast.html, s);
  const aliases = Object.entries(components);

  if (aliases.length === 0) {
    return { code: content };
  }
  const script = compiled.ast.instance || compiled.ast.module;
  const wrappers = aliases
    .map(
      ([alias, expression]) =>
        `const ${alias} = ${prefix}sveltify(${expression}, ${portal}, ${prefix}ReactDOM${renderToString});`,
    )
    .join(";");

  if (!script) {
    s.prepend(`<script>\n${imports.join("; ")}\n\n${wrappers}\n</script>\n\n`);
  } else {
    const program = script.content as any;
    s.appendRight(program.end, wrappers);
    s.appendRight(program.start, `${imports.join("; ")}; `);
  }
  return {
    code: s.toString(),
    map: s.generateMap(),
  };
}

function replaceReactTags(
  node: TemplateNode,
  content: MagicString,
  components: Record<string, string> = {},
) {
  /* eslint-disable no-param-reassign */
  if (node.type === "Element" && node.name.startsWith("react:")) {
    const tag = node as Element;
    const componentExpression = tag.name.slice(6);
    const alias = `React$${componentExpression.replace(/\./g, "$")}`;
    const tagStart = node.start;
    const tagEnd = node.end;
    const closeStart = tagEnd - tag.name.length - 3;
    const hasCloseTag =
      content.slice(closeStart, closeStart + 8) === `</react:`;
    content.overwrite(tagStart + 1, tagStart + 1 + tag.name.length, alias);
    if (hasCloseTag) {
      content.overwrite(
        closeStart + 2,
        closeStart + 2 + tag.name.length,
        alias,
      );
    }
    if (!components[alias]) {
      if (componentExpression.match(/^[a-z-]+$/)) {
        components[alias] = `"${componentExpression}"`;
      } else {
        components[alias] = componentExpression;
      }
    }
    tag.attributes.forEach((attr) => {
      if (attr.type === "EventHandler" && attr.expression !== null) {
        const event = attr as Transition;
        if (event.modifiers.length > 0) {
          throw new Error(
            "event modifier are not (yet) supported for React components",
          );
        }
        const eventStart = event.start;
        content.overwrite(
          eventStart,
          eventStart + 4,
          `on${event.name[0].toUpperCase()}`,
        );
      }
    });
    if (node.children && node.children.length > 0) {
      const isTextContent =
        node.children.filter(
          (child) => ["Text", "MustacheTag"].includes(child.type) === false,
        ).length === 0;
      const escaped: string[] = [];
      if (isTextContent) {
        // Convert text & expresions into a children prop.
        escaped.push('"');
        node.children.forEach((child) => {
          if (child.type === "Text") {
            escaped.push(
              child.data.replace(/"/g, `{'"'}`).replace(/\n/g, `{'\\n'}`),
            );
          } else if (child.type === "MustacheTag") {
            const expression = content.original.slice(child.start, child.end);
            escaped.push(expression);
          } else {
            throw new Error(`Unexpected node type:${child.type}`);
          }
        });
        escaped.push('"');
        // slot was converted to children prop
        content.appendRight(
          node.children[0].start - 1,
          ` children=${escaped.join("")} /`,
        );
        content.remove(node.children[0].start, node.end);
        return components;
      }
    }
  }
  // traverse children & branching blocks
  node.children?.forEach((child) => {
    replaceReactTags(child, content, components);
  });
  node.else?.children?.forEach((child: TemplateNode) => {
    replaceReactTags(child, content, components);
  });
  node.pending?.children?.forEach((child: TemplateNode) => {
    replaceReactTags(child, content, components);
  });
  node.then?.children?.forEach((child: TemplateNode) => {
    replaceReactTags(child, content, components);
  });
  node.catch?.children?.forEach((child: TemplateNode) => {
    replaceReactTags(child, content, components);
  });
  return components;
}
