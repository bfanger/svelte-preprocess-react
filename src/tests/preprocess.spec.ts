import { readFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { vitePreprocess } from "@sveltejs/vite-plugin-svelte";
import { preprocess } from "svelte/compiler";
import { describe, expect, it, vi } from "vitest";
import preprocessReact from "../lib/preprocessReact";

describe("svelte-preprocess-react", () => {
  it("should process <react:component> tags", async () => {
    const filename = resolveFilename("./fixtures/Container.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should process <react:component> tags", async () => {
    const filename = resolveFilename("./fixtures/Multiple.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should import 'react-dom/server' when ssr is enabled", async () => {
    const filename = resolveFilename("./fixtures/Container.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact({ ssr: true }), {
      filename,
    });
    expect(output.code).toMatch("react-dom/server");
    expect(output.code).toMatchSnapshot();
  });

  it("should not import 'react-dom/server' when ssr is disabled", async () => {
    const filename = resolveFilename("./fixtures/Container.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact({ ssr: false }), {
      filename,
    });
    expect(output.code).not.toMatch("react-dom/server");
    expect(output.code).toMatchSnapshot();
  });

  it("should fail on bindings", async () => {
    const filename = resolveFilename("./fixtures/Binding.svelte");
    const src = await readFile(filename, "utf8");
    const consoleMock = vi
      .spyOn(console, "warn")
      .mockImplementation(() => undefined);
    await preprocess(src, preprocessReact(), { filename });
    expect(consoleMock).toBeCalledWith(
      `Two-way binding is not compatible with React components:
  <react.Clicker bind:count> in ${filename} on line 9`,
    );
    consoleMock.mockReset();
  });

  it("should portal slotted content as children", async () => {
    const filename = resolveFilename("./fixtures/Slots.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should inject a script tag", async () => {
    const filename = resolveFilename("./fixtures/NoScript.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toContain("<script>");
    expect(output.code).toMatchSnapshot();
  });
  it("should support typescript when using preprocess", async () => {
    const filename = resolveFilename("./fixtures/Typescript.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: vitePreprocess() }),
      { filename },
    );
    expect(output.code).toMatchSnapshot();
  });
  it("should convert text content to react children prop", async () => {
    const filename = resolveFilename("./fixtures/SlottedText.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: vitePreprocess() }),
      { filename },
    );
    expect(output.code).toMatchSnapshot();
  });
  it("should process <react:Context.Provider> tags", async () => {
    const filename = resolveFilename("./fixtures/Provider.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: vitePreprocess() }),
      { filename },
    );
    expect(output.code).toMatchSnapshot();
  });
  it("should process <react:element> (lowercase) tags", async () => {
    const filename = resolveFilename("./fixtures/Element.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });
  it("should process {:else} {:then} and {:catch} sections", async () => {
    const filename = resolveFilename("./fixtures/Blocks.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: vitePreprocess() }),
      { filename },
    );
    expect(output.code).toMatchSnapshot();
  });
  it("should process on:event forwarding", async () => {
    const filename = resolveFilename("./fixtures/Forwarding.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(
      src,
      preprocessReact({ preprocess: vitePreprocess() }),
      { filename },
    );
    expect(output.code).toMatchSnapshot();
  });

  it("should process <react.Component.Item> tags", async () => {
    const filename = resolveFilename("./fixtures/List.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should process {...rest} props", async () => {
    const filename = resolveFilename("./fixtures/RestProps.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });

  it("should detect tailing comma when adding aliases", async () => {
    const filename = resolveFilename("./fixtures/TrailingComma.svelte");
    const src = await readFile(filename, "utf8");
    const output = await preprocess(src, preprocessReact(), { filename });
    expect(output.code).toMatchSnapshot();
  });
});

const base = dirname(fileURLToPath(import.meta.url));
function resolveFilename(filename: string) {
  return resolve(base, filename);
}
