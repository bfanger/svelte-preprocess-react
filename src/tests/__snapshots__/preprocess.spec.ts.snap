// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svelte-preprocess-react > should convert text content to react children prop 1`] = `
"<script lang="ts">import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  import Alert from "../../demo/react-components/Alert";

  const count = 0;
;const React$Alert = React$$sveltify(Alert, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Alert react$Children="Static Text" />

<React$Alert react$Children="{'\\n'}  {'"'}Multiline content{'"'}. Lorem ipsum dolor sit amet consectetur adipisicing elit.{'\\n'}  Suscipit nisi atque asperiores.{'\\n'}" />

<React$Alert react$Children="Value: {count + 1}" />
"
`;

exports[`svelte-preprocess-react > should import 'react-dom/server' when ssr is enabled 1`] = `
"<script>import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  // @ts-nocheck
  import Clicker from "./Clicker";

  let count = 1;
;const React$Clicker = React$$sveltify(Clicker, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Clicker
  {count}
  onCount={(next) => {
    count = next;
  }}
/>
"
`;

exports[`svelte-preprocess-react > should inject a script tag 1`] = `
"<script>
import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";

const React$Counter = React$$sveltify(Counter, React$$createPortal, React$$ReactDOM, React$$renderToString);
</script>

<!-- Counter could be a global variable -->
<React$Counter />
"
`;

exports[`svelte-preprocess-react > should not import 'react-dom/server' when ssr is disabled 1`] = `
"<script>import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom";
  // @ts-nocheck
  import Clicker from "./Clicker";

  let count = 1;
;const React$Clicker = React$$sveltify(Clicker, React$$createPortal, React$$ReactDOM);</script>

<React$Clicker
  {count}
  onCount={(next) => {
    count = next;
  }}
/>
"
`;

exports[`svelte-preprocess-react > should portal slotted content as children 1`] = `
"<script>import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  import Alert from "../../demo/react-components/Alert";
;const React$Alert = React$$sveltify(Alert, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Alert react$Children="A simple primary alertâ€”check it out!" />
"
`;

exports[`svelte-preprocess-react > should process <react:Context.Provider> tags 1`] = `
"<script lang="ts">import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  import type { Context as ContentType } from "react";
  import { used } from "svelte-preprocess-react";

  export let Context: ContentType<any>;
  $: Provider = Context.Provider;

  used(Provider);
;const React$Provider = React$$sveltify(Provider, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Provider react$Children="content" />

<React$Provider>
  <button>ok</button>
</React$Provider>
"
`;

exports[`svelte-preprocess-react > should process <react:component> tags 1`] = `
"<script>import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  // @ts-nocheck
  import Clicker from "./Clicker";

  let count = 1;
;const React$Clicker = React$$sveltify(Clicker, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Clicker
  {count}
  onCount={(next) => {
    count = next;
  }}
/>
"
`;

exports[`svelte-preprocess-react > should process <react:component> tags 2`] = `
"<script>import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  // @ts-nocheck
  import Clicker from "./Clicker";

  let count = 1;
;const React$Clicker = React$$sveltify(Clicker, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<h2>prop and event</h2>
<React$Clicker
  {count}
  onCount={(next) => {
    count = next;
  }}
/>
<h2>prop and Prop event</h2>
<React$Clicker
  count={10}
  onCount={() => {
    console.info("count");
  }}
/>
<h2>prop</h2>
<React$Clicker {count} />
<h2>.</h2>
<React$Clicker />
"
`;

exports[`svelte-preprocess-react > should process <react:element> (lowercase) tags 1`] = `
"<script>
import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";

const React$button = React$$sveltify("button", React$$createPortal, React$$ReactDOM, React$$renderToString);
</script>

<React$button onclick={() => console.info("clicked")}>
  <slot />
</React$button>
"
`;

exports[`svelte-preprocess-react > should process {:else} {:then} and {:catch} sections 1`] = `
"<script lang="ts">import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  import type React from "react";

  const number = 1;

  const Component: React.FC = () => null;
;const React$Component = React$$sveltify(Component, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<!-- eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -->
{#if number === 1}
  <React$Component />
  <!-- eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -->
{:else if number === 2}
  <React$Component />
{:else}
  <React$Component />
{/if}

{#each [] as item}
  <React$Component />
{:else}
  <React$Component />
{/each}

{#await Promise.resolve()}
  <React$Component />
{:then}
  <React$Component />
{:catch}
  <React$Component />
{/await}
"
`;

exports[`svelte-preprocess-react > should process on:event forwarding 1`] = `
"<script lang="ts">import { sveltify as React$$sveltify } from "svelte-preprocess-react"; import React$$ReactDOM from "react-dom/client"; import { createPortal as React$$createPortal} from "react-dom"; import { renderToString as React$$renderToString } from "react-dom/server";
  import { used } from "svelte-preprocess-react";
  import Clicker from "./Clicker";

  used(Clicker);

  let { value = 0, onCount } = $props<{
    value: number;
    onCount: (count: number) => void;
  }>();
;const React$Clicker = React$$sveltify(Clicker, React$$createPortal, React$$ReactDOM, React$$renderToString);</script>

<React$Clicker count={value} {onCount} />
"
`;

exports[`svelte-preprocess-react > should support typescript when using preprocess 1`] = `
"<script lang="ts">
  export let title: string;
</script>

<h1>{title}</h1>
"
`;
